### DISCUSSION TITLE
Using sento.future:with-fut without blocking?

### DISCUSSION THREAD

**Discussion by swapneils** [mdbergmann/cl-gserver#71] [Q&A]
I've started playing around with this library, and noticed that whenever I try to make a future using `with-fut` or `with-fut-resolve`, the call to `resolve-fun` in the macroexpanded code blocks on the calculation in the future, which seems to make _creating_ a future meaningless. This is the case even when `fresolve` is called in a thread in `with-fut-resolve`, since `fresolve` expands to a `make-future` call instead of isolating the call to the `resolve-fun`.

If we just use `sento.future:make-future` directly, we can wrap the `resolve-fun` call in a thread to produce a non-blocking future, as mentioned in the function's docstring. Example:
```
(let ((fut (make-future (lambda (execute-fun) 
                                (let ((promise (+ 1 2)))
                                  (bt:make-thread (lambda ()
                                                    (sleep 0.5)
                                                    (funcall execute-fun promise))))))))
        (print fut) ;; => #<FUTURE promise: #<PROMISE finished: NIL errored: NIL forward: NIL {1002B47DE3}>> 
        (sleep 1)
        (print fut)) ;; => #<FUTURE promise: #<PROMISE finished: T errored: NIL forward: NIL {1002B47DE3}>> 
```

Is there any way to make `with-fut` and `with-fut-resolve` similarly non-blocking? And if not, what is the intended use-case for those macros?

**Comment by mdbergmann** [comment]
Hello.

As far as I remember. The futures in this package are agnostic to how the computations are performed. So basically it's up to the user to whether it is needed to fulfill the future asynchronously or not. For asynchronous fulfillment the user can choose to spawn a new thread, or use Sento.tasks library (preferred) or somehow else. The mapping of futures as ca; be seen in the tests package works in either case.

The ˋwith-futˋ macros are just for convenience and for nicer syntax.


Regards
